qualtran/bloqs/mod_arithmetic/mod_multiplication.py

```diff
@@ -265,3 +273,463 @@ def _modmul_symb() -> CModMulK:
+                [Shaped((2**self.window_size,))],
+                selection_bitsizes=(self.window_size,),
+                target_bitsizes=(self.window_size + self.bitsize,),
+                log_block_sizes=log_block_sizes,
+            )
+        inv_mod = pow(self.mod, 2 ** (self.window_size - 1) - 1, 2**self.window_size)
+        N = 2**self.window_size
+        data = (-np.arange(N) * inv_mod) % N
+        data *= self.mod
+        return QROAMClean(
+            [data],
+            selection_bitsizes=(self.window_size,),
+            target_bitsizes=(self.window_size + self.bitsize,),
+        )
+
+    def on_classical_vals(self, x: Sequence[int], y: int, target: Sequence[int], qrom_index: int):
+        if is_symbolic(self.bitsize) or is_symbolic(self.window_size):
+            raise ValueError(f'classical action is not supported for {self}')
+        dtype = QMontgomeryUInt(self.window_size + self.bitsize)
+        target_val = QMontgomeryUInt.from_bits(dtype, target)
+        for i in range(self.window_size):
+            if x[i]:
+                target_val += y << i
+        qrom_index = target_val & (2**self.window_size - 1)
+        Tm = self.qrom.data[0][qrom_index]
+        target_val = (target_val + Tm) >> self.window_size
+        target = QMontgomeryUInt.to_bits(dtype, target_val)
+        return {'target': target, 'qrom_index': qrom_index, 'x': x, 'y': y}
+
+    def build_composite_bloq(
+        self, bb: 'BloqBuilder', x: NDArray[Soquet], y: Soquet, target: NDArray[Soquet], qrom_index: Soquet  # type: ignore[type-var]
+    ):
+        if is_symbolic(self.window_size):
+            raise DecomposeNotImplementedError(f'symbolic decomposition not supported for {self}')
+        for i in range(self.window_size):
+            z = bb.join(target[-self.bitsize - 1 - i : len(target) - i])
+            x[i], y, z = bb.add(
+                CAdd(QMontgomeryUInt(self.bitsize), QMontgomeryUInt(self.bitsize + 1)),
+                ctrl=x[i],
+                a=y,
+                b=z,
+            )
+            z_arr = bb.split(z)
+            target[-self.bitsize - 1 - i : len(target) - i] = z_arr
+
+        m = bb.join(target[-self.window_size :], QMontgomeryUInt(self.window_size))
+        m, qrom_index = bb.add(Xor(QMontgomeryUInt(self.window_size)), x=m, y=qrom_index)
+        target[-self.window_size :] = bb.split(m)
+
+        qrom_index, qrom_target = bb.add(self.qrom, selection=qrom_index)
+        z = bb.join(target)
+        qrom_target, z = bb.add(
+            Add(QMontgomeryUInt(self.bitsize + self.window_size)), a=qrom_target, b=z
+        )
+        qrom_index = bb.add(self.qrom.adjoint(), selection=qrom_index, target0_=qrom_target)
+        target_arr = bb.split(z)
+        target_arr = np.roll(target_arr, self.window_size)
+
+        return {'x': x, 'y': y, 'target': target_arr, 'qrom_index': qrom_index}
+
+    def build_call_graph(self, ssa: 'SympySymbolAllocator') -> Set['BloqCountT']:
```

Comment:
we've changed this to return a dictionary now. Can you update? Thank you!

Reply:
done